src/state/presence.ts

```diff
@@ -0,0 +1,38 @@
+import { useCloudReducer } from "@compose-run/client";
+import appName from "./appName";
+import { useState, useEffect } from "react";
+
+// Maps user id to "last seen" unix timestamp.
+type PresenceDB = { [userId: number]: number };
+
+export type PresenceAction = number;
+type PresenceActionError = string;
+
+export const usePresence = () => {
+  const [presenceDb, dispatchPresence] = useCloudReducer<
+    PresenceDB,
+    PresenceAction,
+    PresenceActionError
+  >({
+    name: `${appName}/user-presence`,
+    initialState: {},
+    reducer: (users, lastSeen, { userId }) => {
+      users[userId] = lastSeen;
+      return users;
+    },
+  });
+  const [numberOnline, setNumberOnline] = useState(0);
+  useEffect(() => {
+    setInterval(() => {
+      const nw = Date.now();
+      dispatchPresence(nw);
+      console.log(presenceDb);
+      setNumberOnline(
+        Object.values(presenceDb || {}).filter(
+          (lastSeen: number) => lastSeen > nw - 30 * 1000
+        ).length
+      );
+    }, 1000 * 20);
+  });
+  return numberOnline;
```

Comment:
```suggestion
  useEffect(() => {
    // dispatch presence on first load, and every 20 seconds thereafter
    dispatchPresence(Date.now());
    const interval = setInterval(() => {
      dispatchPresence(Date.now());        
    }, 1000 * 20);
    return () => clearInterval(interval)
  }, [userId]); // only do this effect when the userId changes
  return Object.values(presenceDb).filter(
      (lastSeen: number) => lastSeen > nw - 30 * 1000
    ).length
  );
```

Reply:
Do you see why we calculate the `numberOnline` value outside of the interval? (it's a reactive value)

The code I wrote above requires that you pull in the `useUser` hook in this function

Reply:
I think I get it!